### PR TITLE
Migrate hyper to reqwest and axum, update dependencies

### PR DESCRIPTION
**Pull request by threema-donat**
- Upgrade all dependencies
- Bump rust toolchain to 1.77.2
- Replace mockito with wiremock
- Migrate from hyper to reqwest and axum
- Switch from log to tracing

### CODE REVIEW

File: src/errors.rs [hunk 1]
```diff
@@ -51,17 +54,25 @@ pub enum InfluxdbError {
...
 
-#[derive(Debug, Clone, PartialEq)]
-pub struct ServiceError(String);
+#[derive(Debug)]
+pub struct ServiceError(anyhow::Error);
 
-impl fmt::Display for ServiceError {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "ServiceError: {}", self.0)
+// Tell axum how to convert `AppError` into a response.
```
Comment (threema-danilo) [inline src/errors.rs:60]: I guess you meant `ServiceError`, not `AppError`?

File: src/push/fcm.rs [hunk 2]
```diff
@@ -68,10 +59,25 @@ pub struct MessageResult {
     pub error: Option<String>,
 }
 
+#[derive(Debug)]
+pub struct FcmStateConfig {
```
Comment (threema-danilo) [inline src/push/fcm.rs:63]: What is a "state config"? From looking at the usages, this seems to be the API key plus the endpoint. None of that is stateful. Why not just `FcmConfig` (and `HmsConfig` etc)? Or maybe `FcmEndpointConfig`?

File: src/push/hms.rs [hunk 3]
```diff
@@ -32,32 +34,45 @@ use crate::{
...
-// Server endpoints
-#[cfg(not(test))]
-fn hms_endpoint(endpoint_type: EndpointType) -> &'static str {
-    match endpoint_type {
-        EndpointType::Login => "https://oauth-login.cloud.huawei.com",
-        EndpointType::Push => "https://push-api.cloud.huawei.com",
+type SharedHmsConfig = Arc<HmsStateConfig>;
+
+impl HmsStateConfig {
+    pub fn new() -> SharedHmsConfig {
```
Comment (threema-danilo) [inline src/push/hms.rs:45]: The `FcmStateConfig` has a `new_shared` method, while here we just use `new`. Maybe the two could be made consistent? (I think `new_shared` is pretty good naming, if the returned type is wrapped in an Arc.)

File: src/push/hms.rs [hunk 4]
```diff
@@ -518,50 +553,50 @@ mod tests {
                 },
             );
 
-            // Mock HMS auth endpoint
-            let m = mock("POST", "/oauth2/v3/token")
-                .with_body(format!(
+            let mock_server = MockServer::start().await;
```
Comment (threema-danilo) [inline src/push/hms.rs:556]: Nice, wiremock looks pretty cool 🙂

File: src/server.rs [hunk 5]
```diff
@@ -127,113 +130,64 @@ pub async fn serve(
...
-    apns_client_sbox: Arc<Mutex<ApnsClient>>,
-    hms_contexts: Arc<HashMap<String, HmsContext>>,
-    threema_gateway_client: HttpClient,
-    threema_gateway_config: Option<ThreemaGatewayConfig>,
-    threema_gateway_private_key: Option<ThreemaGatewayPrivateKey>,
-    influxdb: Option<Arc<Influxdb>>,
+fn get_router(state: AppState) -> Router {
+    axum::Router::new()
+        .route(PUSH_PATH, post(handle_push_request))
+        .with_state(state)
```
[the remark this answers is not in the document]
Comment (threema-donat) [reply src/server.rs:189]: Exactly, that's how it works :)

File: src/push/fcm.rs (hunk 2 above)
Comment (threema-donat) [reply src/push/fcm.rs:63]: I called it this way to differentiate it from the config that is parsed from the config file, and I wasn't happy with the naming, too... FcmConfig is fine for me if you don't mind the same name

Comment (threema-donat) [reply src/push/fcm.rs:63]: renamed to FcmConfig

File: src/push/hms.rs (hunk 3 above)
Comment (threema-donat) [reply src/push/hms.rs]: Renamed the method for consistency

File: src/errors.rs [hunk 6]
```diff
@@ -11,11 +11,14 @@ pub enum PushRelayError {
     #[error("APNs error: {0}")]
     Apns(#[from] A2Error),
 
-    #[error("Hyper error: {0}")]
-    Hyper(#[from] HyperError),
-
     #[error("Reqwest error: {0}")]
     Reqwest(#[from] ReqwestError),
+
+    #[error("{reason}: {source}")]
```
[the remark this answers is not in the document]
Comment (threema-donat) [reply src/errors.rs]: Added prefix to message

File: src/influxdb.rs [hunk 7]
```diff
@@ -96,20 +91,19 @@ impl Influxdb {
...
 
-    async fn log(&self, body: Body) -> InfluxdbResult {
-        // format URL
-        let uri: Uri = format!("{}/write?db={}", &self.connection_string, &self.db)
-            .parse()
-            .unwrap();
+    async fn log<B>(&self, body: B) -> InfluxdbResult
+    where
+        B: Into<Body>,
+    {
```
[the remark this answers is not in the document]
Comment (threema-donat) [reply src/influxdb.rs]: Applied suggestion

File: src/http_client.rs [hunk 8]
```diff
@@ -8,6 +8,19 @@ use hyper_rustls::{HttpsConnector, HttpsConnectorBuilder};
...
 
+/// Create a HTTP 1 client instance.
+///
+/// Parameter: Timeout for idle sockets being kept-alive
+pub fn make_client_v2(pool_idle_timeout_seconds: u64) -> Result<reqwest::Client, reqwest::Error> {
+    reqwest::Client::builder()
+        .pool_idle_timeout(Duration::from_secs(pool_idle_timeout_seconds))
+        .http1_only()
+        .use_rustls_tls()
+        .https_only(false)
```
[the remark this answers is not in the document]
Comment (threema-donat) [reply src/http_client.rs:13]: Added parameter

File: src/http_client.rs [hunk 9]
```diff
@@ -8,6 +8,19 @@ use hyper_rustls::{HttpsConnector, HttpsConnectorBuilder};
...
+        .tls_built_in_root_certs(true)
```
[the remark this answers is not in the document]
Comment (threema-donat) [reply src/http_client.rs]: true, fixed!

File: src/errors.rs [hunk 10]
```diff
@@ -51,17 +54,25 @@ pub enum InfluxdbError {
...
 
-impl fmt::Display for ServiceError {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "ServiceError: {}", self.0)
+// Tell axum how to convert `AppError` into a response.
+impl IntoResponse for ServiceError {
+    fn into_response(self) -> Response {
+        (
+            StatusCode::INTERNAL_SERVER_ERROR,
+            format!("Something went wrong: {}", self.0),
```
[the remark this answers is not in the document]
Comment (threema-donat) [reply src/errors.rs]: this was actually never used, but it still made sense to mitigate some duplication with the error

File: src/errors.rs (hunk 1 above)
Comment (threema-donat) [reply src/errors.rs]: Right!

**Review by threema-danilo — APPROVED** [review]

**Pull request merged**